SharedCryptoPkg/Library/CryptLibSharedDriver/Common/CommonSharedCryptoLib.c

```diff
@@ -0,0 +1,3884 @@
+/** @file
+SharedCryptoLib.c
+
+This defines the implementation of BaseCryptLib and calls out to the protocol
+published by SharedCryptoDriver. The ProtocolFunctionNotFound and GetProtocol
+funtion is defined by the flavor of the library.
+
+Copyright (c) 2019, Microsoft Corporation
+
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+**/
+
+#include <Library/BaseCryptLib.h>
+#include <Library/BaseLib.h>
+#include <Library/DebugLib.h>
+#include <SharedCryptoHelpers.h>
+
+
+//=====================================================================================
+//    One-Way Cryptographic Hash Primitives
+//=====================================================================================
+
+/**
+  Retrieves the size, in bytes, of the context buffer required for MD4 hash operations.
+
+  If this interface is not supported, then return zero.
+
+  @return  The size, in bytes, of the context buffer required for MD4 hash operations.
+  @retval  0   This interface is not supported.
+
+**/
+UINTN
+EFIAPI
+Md4GetContextSize (
+    VOID)
+{
+  SHARED_CRYPTO_FUNCTIONS *prot = GetProtocol();
```

Comment:
Instead of looking up the protocol on every API, it would be better if the library has a CONSTRUCTOR that looks up the protocol and stores it in a module global and the lib instance has a [Depex] on the protocol GUID.  This will work for DXE and SMM instances.  If the PEI instance is in an XIP module, then the module global will not work, so you may need to use this style of the PEI use case.

Reply:
That's sort of what we do. For DXE and SMM, we store the found pointer and don't do a LocateProtocol if we have already found it before. There is a DEPEX for these libraries on their respective protocols so DXE modules that rely on BaseCryptLib won't load if SharedCrypto doesn't.

Reply:
I'll look into using a constructor instead.